core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,663 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class for classification algorithms."""
+
+import collections
+import keyword
+import StringIO
+import token
+import tokenize
+
+from core.classifiers import base
+
+import numpy as np
+from sklearn.feature_extraction import text as sklearn_text
+from sklearn import model_selection
+from sklearn import svm
+
+# Token which will be used to replace variable and method name when tokenizing
+# program.
+VAR_TOKEN = 'V'
+
+# Token which will be used to replace unkownn token when tokenizing
+# program.
+UNK_TOKEN = 'UNK'
+
+# T is minimum length of a substring that winnowing should match when
+# comparing two programs.
+# Minimum value of T. T is model parameter of winnowing algorithm.
+T_MIN = 3
+# Maximum value of T. T is model parameter of winnowing algorithm.
+T_MAX = 11
+
+# K is length of substring for which hash value is to be generated. Each
+# substring of length K in program will be converted to equivalent hash
+# value (K <= T).
+# Minimum value of K (K <= T). K is model parameter of winnowing algorithm.
+K_MIN = 3
+
+def get_token(program):
+    """Generate tokens for program using tokenize module.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+
+    Yields:
+        int. Integer value which represents type of token.
+        str. Token.
+    """
+    for token_id, token_name, _, _, _ in tokenize.generate_tokens(
+            StringIO.StringIO(program).readline):
+        yield (token_id, token_name)
+
+
+def cv_tokenizer(program, var_token=VAR_TOKEN):
+    """Custom tokenizer of Python program for CountVectorizer in sci-kit.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+        var_token: str. Replace variable and method names with this string when
+            tokenizing data.
+
+    Returns:
+        list(str). A list containing tokens of input program.
+    """
+    token_program = []
+    for token_id, token_name in get_token(program):
+        # Ignore all newline, comment and empty string tokens.
+        if (token_id == token.N_TOKENS or token_id == 54
+                or token_name.strip() == ''):
+            continue
+        elif token_id == token.NAME:
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword. Treat all variables and methods same.
+            if token_name in keyword.kwlist:
+                token_program.append(token_name)
+            else:
+                token_program.append(var_token)
+        else:
+            token_program.append(token_name)
+
+    return token_program
+
+
+def generate_token_to_id(
+        data, threshold, var_token=VAR_TOKEN, unk_token=UNK_TOKEN):
+    """Generates a list of valid tokens and assigns a unique ID to each
+    token.
+
+    Args:
+        data: dict. A dictionary containing training_data.
```

Comment:
If the structure is different, then, it makes more sense to indicate its structure'. I'm saying this because after looking at `data`, I assume that it has the same structure as the training_data you receive from Oppia.